engine/table/src/main/java/io/deephaven/engine/table/impl/UngroupOperation.java

```diff
@@ -431,214 +332,420 @@ private class UngroupListener extends BaseTable.ListenerImpl {
+                            break SHIFT_LOOP;
                         }
+                        // we successfully advanced
+                        nextResultKey = rsit.currentRangeStart();
+                        nextSourceKey = nextResultKey >> requiredBase;
+
+                        if (nextResultKey > oldRangeEnd) {
+                            // no need to shift things that don't exist anymore
+                            continue;
+                        }
+
+                        // get the range from our result rowset
+                        Assert.eq(oldRangeStart, "oldRangeStart", rsit.currentRangeStart(),
+                                "rsit.currentRangeStart()");
+                        final long actualRangeEnd = Math.min(rsit.currentRangeEnd(), oldRangeEnd);
+                        shiftBuilder.shiftRange(oldRangeStart, actualRangeEnd, resultShiftAmount);
+                        removedByShiftBuilder.addRange(oldRangeStart, actualRangeEnd);
+                        addedByShiftBuilder.addRange(oldRangeStart + resultShiftAmount,
+                                actualRangeEnd + resultShiftAmount);
+
                     }
                 }
             }
+        }
 
-            resultRowset.remove(removedByShiftBuilder.build());
-            resultRowset.insert(addedByShiftBuilder.build());
-            resultRowset.insert(addedRowSet);
-            for (final ColumnSource<?> source : resultMap.values()) {
-                if (source instanceof UngroupedColumnSource) {
-                    ((UngroupedColumnSource<?>) source).setBase(newBase);
+        private void rebase(final int newBase, final TableUpdate upstream, final long[] modifiedCurrentSizes) {
+            final RowSetBuilderSequential addedBuilder = RowSetFactory.builderSequential();
+            final RowSetBuilderSequential removedBuilder = RowSetFactory.builderSequential();
+
+            final RowSetBuilderSequential addedByModifiesBuilder = RowSetFactory.builderSequential();
+            final RowSetBuilderSequential removedByModifiesBuilder = RowSetFactory.builderSequential();
+            final RowSetBuilderSequential modifiedByModifiesBuilder = RowSetFactory.builderSequential();
+
+            final int computedAddBase = evaluateAdded(upstream.added(), addedBuilder, newBase, false);
+            final int computedModifiedBase = evaluateModified(upstream.modified(), upstream.getModifiedPreShift(),
+                    addedByModifiesBuilder, removedByModifiesBuilder, modifiedByModifiesBuilder, newBase,
+                    new MutableObject<>(modifiedCurrentSizes));
+            Assert.leq(computedAddBase, "computedAddBase", newBase, "newBase");
+            Assert.leq(computedModifiedBase, "computedAddBase", newBase, "newBase");
+
+            evaluateRemoved(upstream.removed(), removedBuilder);
+
+            final WritableRowSet removedRowSet = removedBuilder.build();
+            removedRowSet.insert(removedByModifiesBuilder.build());
```

Comment:
Be sure to close the result of `removedByModifiesBuilder.build()` after the insert. You should check for similar issues in the happy path.